ui/src/shared/components/AnnotationFilterControlInput.tsx

```diff
@@ -0,0 +1,157 @@
+  filteredSuggestions: string[]
+}
+
+const NUM_SUGGESTIONS = 8
+
+class AnnotationFilterControlInput extends PureComponent<Props, State> {
+  public static getDerivedStateFromProps(props) {
+    const {suggestions, value} = props
+
+    const filtered: string[] = suggestions.filter(
+      (v, i) => v.includes(value) && i < NUM_SUGGESTIONS
+    )
+    const unique: string[] = [...new Set(filtered)]
+
+    unique.sort((a, b) => a.localeCompare(b))
+
+    return {filteredSuggestions: unique}
+  }
+
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      filteredSuggestions: props.suggestion,
+      selectionIndex: 0,
+      areSuggestionsVisible: false,
+    }
+  }
+
+  public render() {
+    const {value} = this.props
+
+    return (
+      <ClickOutside onClickOutside={this.handleClickOutside}>
+        <div className="suggestion-input">
+          <input
+            className={this.inputClass}
+            value={value}
+            onChange={this.handleChange}
+            onFocus={this.handleFocus}
+            onKeyUp={this.handleKeyUp}
+          />
+          <div
+            className={`suggestion-input--suggestions dropdown-menu ${
+              this.suggestionsClass
+            }`}
+          >
+            {this.renderSuggestions()}
+          </div>
+        </div>
+      </ClickOutside>
+    )
+  }
+
+  private renderSuggestions() {
+    const {onSelect} = this.props
+    const {selectionIndex, filteredSuggestions} = this.state
+
+    return filteredSuggestions.map((v, i) => {
+      const selectedClass = i === selectionIndex ? 'selected' : ''
+      const onClick = () => {
+        onSelect(v)
+        this.setState({areSuggestionsVisible: false})
+      }
+
+      return (
+        <div
+          key={v}
+          className={`suggestion-input-suggestion dropdown-item ${selectedClass}`}
+          onClick={onClick}
+        >
+          {v}
+        </div>
+      )
+    })
+  }
+
+  private handleChange = (e: ChangeEvent<HTMLInputElement>) => {
+    this.props.onChange(e.target.value)
+  }
+
+  private handleFocus = () => {
+    this.setState({areSuggestionsVisible: true})
+
+    this.props.onFocus()
+  }
+
+  private handleClickOutside = () => {
+    this.setState({areSuggestionsVisible: false})
+  }
+
+  private handleKeyUp = (e: KeyboardEvent<HTMLInputElement>) => {
+    const {key, ctrlKey} = e
+
+    if (key === 'Enter') {
+      const {onSelect} = this.props
+      const {selectionIndex, filteredSuggestions} = this.state
+
+      e.currentTarget.blur()
+      onSelect(filteredSuggestions[selectionIndex])
+      this.setState({areSuggestionsVisible: false})
+    } else if (
+      key === 'ArrowDown' ||
+      (ctrlKey && (key === 'j' || key === 'n'))
+    ) {
+      this.incrementSelectionIndex(1)
+    } else if (key === 'ArrowUp' || (ctrlKey && (key === 'k' || key === 'p'))) {
+      this.incrementSelectionIndex(-1)
+    }
+  }
+
+  private incrementSelectionIndex = (di: number): void => {
```

Comment:
could you clarify this param `di`? i don't know what that means when i read this contract

Reply:
It's a d3 idiom, read “delta i", as in a “small change of i” (originally inspired by [this](https://en.wikipedia.org/wiki/Integral#Meaning_of_the_symbol_dx) I believe)